File: examples/runnable/10-child-workflows/go/batch.go
// Package main holds the parent/child lab. The Go equivalent of the Java
// child-workflows lab. The parent starts one child per item - each with its own
// stable Workflow ID, so it is separately queryable / signalable / cancelable -
// then waits for all of them.
package main

import (
	"strings"
	"time"

	"go.temporal.io/sdk/workflow"
)

const TaskQueue = "child-workflows"

// ItemWorkflow is a child Workflow - its own Workflow ID, its own history,
// separately addressable.
func ItemWorkflow(ctx workflow.Context, item string) (string, error) {
	// A durable sleep so each child is visibly "running" in the Web UI for a moment.
	if err := workflow.Sleep(ctx, 500*time.Millisecond); err != nil {
		return "", err
	}
	return "processed[" + item + "] in child " + workflow.GetInfo(ctx).WorkflowExecution.ID, nil
}

// BatchWorkflow starts one child per item in parallel, then collects the results.
func BatchWorkflow(ctx workflow.Context, items []string) (string, error) {
	futures := make([]workflow.ChildWorkflowFuture, 0, len(items))
	for _, item := range items {
		// Each child gets a stable, independent Workflow ID.
		childCtx := workflow.WithChildOptions(ctx, workflow.ChildWorkflowOptions{
			WorkflowID: "item-" + item,
		})
		futures = append(futures, workflow.ExecuteChildWorkflow(childCtx, ItemWorkflow, item))
	}

	results := make([]string, 0, len(items))
	for _, f := range futures {
		var result string
		if err := f.Get(ctx, &result); err != nil {
			return "", err
		}
		results = append(results, result)
	}
	return strings.Join(results, "\n"), nil
}

File: examples/runnable/10-child-workflows/python/batch.py
"""Parent/child Workflows. The Python equivalent of the Java child-workflows lab.

The parent starts one child per item - each with its own stable Workflow ID, so
it is separately queryable / signalable / cancelable - then waits for all of them.
"""

import asyncio
from datetime import timedelta

from temporalio import workflow

TASK_QUEUE = "child-workflows"


@workflow.defn
class ItemWorkflow:
    """A child Workflow - its own Workflow ID, its own history, separately addressable."""

    @workflow.run
    async def process_item(self, item: str) -> str:
        # A durable sleep so each child is visibly "running" in the Web UI for a moment.
        await workflow.sleep(timedelta(milliseconds=500))
        return f"processed[{item}] in child {workflow.info().workflow_id}"


@workflow.defn
class BatchWorkflow:
    @workflow.run
    async def run(self, items: list[str]) -> str:
        # Start every child before awaiting any → they run in parallel. Each child
        # gets a stable, independent Workflow ID.
        handles = [
            await workflow.start_child_workflow(
                ItemWorkflow.process_item,
                item,
                id=f"item-{item}",
            )
            for item in items
        ]
        results = await asyncio.gather(*handles)
        return "\n".join(results)
